### PR TITLE
Fix LightningCLI signature parameter resolving for some lightning classes

### LINKED ISSUE #11574: DDPPlugin does not accept find_unused_parameters when used with Lightning CLI
## 🐛 Bug

Since DDPPlugin does not have type annotation for `find_unused_parameters` in its `__init__()`, Lightning CLI cannot parse the config (or more correctly, `jsonargparse`?). The following will fail:

```yaml
trainer:
  gpus: 2
  strategy:
    class_path: pytorch_lightning.plugins.DDPPlugin
    init_args:
      find_unused_parameters: false
```

I looked at the new `strategies` module, and it seems like it will have the same problem also.

Right now I'm using a workaround as suggested [here](https://github.com/PyTorchLightning/pytorch-lightning/issues/8262)

```yaml
trainer:
  gpus: 2
  strategy: ddp_find_unused_parameters_false
```

### To Reproduce

```python
# train.py
import torch
from torch.utils.data import DataLoader, Dataset

from pytorch_lightning import LightningModule
from pytorch_lightning.utilities.cli import LightningCLI


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def validation_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("valid_loss", loss)

    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=2)

    def val_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=2)

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


if __name__ == "__main__":
    cli = LightningCLI(BoringModel)
```

```yaml
# config.yaml
trainer:
  gpus: 2
  strategy:
    class_path: pytorch_lightning.plugins.DDPPlugin
    init_args:
      find_unused_parameters: false
```

```bash
python train.py fit --config config.yaml
```

### Expected behavior

Using `find_unused_parameters: false` should work with Lightning CLI config file. This can probably be fixed by adding `find_unused_parameters: Optional[bool] = True` in DDPPlugin/DDPStrategy `__init__()`?

### Environment

- PyTorch Lightning Version (e.g., 1.5.0): 1.5.9
- PyTorch Version (e.g., 1.10): 1.10.1
- Python version (e.g., 3.9): 3.8
- OS (e.g., Linux): Linux
- CUDA/cuDNN version: 11.3.1 / 8.2.0
- GPU models and configuration: 4x 3090
- How you installed PyTorch (`conda`, `pip`, source): conda
- If compiling from source, the output of `torch.__config__.show()`:
- Any other relevant information:

### Additional context

cc @carmocca @mauvilsa

### LINKED ISSUE #11653: Refactor use of **kwargs in PL classes for better LightningCLI support
## Proposed refactor

All the `__init__` parameters from classes in the pytorch-lightning repo that are susceptible to be used via `LightningCLI` should be resolvable by jsonargparse: parameter name, type and docstring description.

A list of potential classes that would be good to refactor is the following. Probably each class needs to be considered separately and some might have higher priority than others.

- [CometLogger](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/comet.py)
- [NeptuneLogger](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/neptune.py)
- [TensorBoardLogger](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/tensorboard.py)
- [WandbLogger](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/loggers/wandb.py)
- [DDPStrategy](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/strategies/ddp.py)
- [DDPSpawnStrategy](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/strategies/ddp_spawn.py)
- [PyTorchProfiler](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/profiler/pytorch.py)
- ...

See Possibilities section below for ideas on how to refactor.

### Motivation

This has been a recurrent issue, for example:

- https://github.com/PyTorchLightning/pytorch-lightning/discussions/8561
- https://github.com/PyTorchLightning/pytorch-lightning/issues/11574
- https://github.com/PyTorchLightning/pytorch-lightning/issues/9887#issuecomment-1018532746
- https://github.com/PyTorchLightning/pytorch-lightning/issues/8008#issuecomment-945540010

There is a proposal to allow specifying values for unresolved parameters which would skip validation https://github.com/omni-us/jsonargparse/issues/114. The downside is no support for several nice features (see Pitch). This can be a solution for some cases. But it is not be the ideal solution, particularly for classes in the pytorch-lightning repo which can be more easily considered for modifications.

### Pitch

With this refactor all features from the CLIs become available:

- Validation of values based on parameter types and early error messages.
- Description of parameters in the help of CLIs.
- Inclusion in reference config files obtained via `--print_config`.
- Linking of arguments. 

### Possibilities

1. Add all parameters to the class. This is how several people have been solving this issue, e.g. https://github.com/PyTorchLightning/pytorch-lightning/issues/9887#issuecomment-1018532746

    - Advantages: Very simple solution. All Pitch features available. No need for deprecations or change how the classes are used.
    - Disadvantages: Duplication of code. The parameters to where these `**kwargs` might change, thus creating problems.

2. Have a specific parameter with dict type. For example in the case of `TensorBoardLogger` there could be a parameter `summary_writer_kwargs: Dict[str, Any]`. The existing `**kwargs` would be deprecated, showing a warning if used.

    - Advantages: Also it is more clear what this parameter is for, unlike a generic `**kwargs`.
    - Disadvantages: No support for features listed in Pitch. Introduces deprecation. Need to change existing code, e.g. from `param1=val, ...` to `summary_writer_kwargs=dict(param1=val, ...)`.

3. Have a specific parameter with a dataclass type. For example in the case of `TensorBoardLogger` there could be a parameter `summary_writer_options: SummaryWriterOptions`. The dataclass would be automatically generated from a signature, thus avoiding code duplication or discrepancies in parameters. See below possible function that could create these dataclasses.

    - Advantages: All Pitch features available.
    - Disadvantages: Introduces deprecation. Need to change existing code, e.g. from `param1=val, ...` to first have an import `from ... import SummaryWriterOptions` and then when used ... summary_writer_kwargs=SummaryWriterOptions(param1=val, ...)`.

    ```python
    import inspect
    from typing import Any, Set
    from dataclasses import make_dataclass, field

    def dataclass_from_signature(name: str, func, skip: Set = None):
        """Creates a dataclass using as fields parameters from the signature of some callable.

        Args:
            name: How the new class should be called, i.e. value to set in ``__name__``.
            func: The callable from which to get the signature.
            skip: Names of parameters to exclude from the signature.

        Example:
            ``MyData = dataclass_from_signature('MyData', MyClass.__init__, skip={'self', 'paramX'})``
        """
        fields = []
        params = list(inspect.signature(func).parameters.values())
        for param in params:
            if skip and param.name in skip:
                continue
            field_info = [param.name, Any if param._annotation == inspect._empty else param._annotation]
            if param.default != inspect._empty:
                field_info.append(field(default=param.default))
            fields.append(tuple(field_info))
        new_dataclass = make_dataclass(name, fields)
        # TODO: Add docstring with description of fields.
        return new_dataclass
    ```

4. Similar to (3) but instead of a dataclass, it would be a normal class also automatically generated. The `__init__` would define the parameters, but it would ignore them. This class would be used as a parent class, so jsonargparse would resolve the parameters via inheritance. For example for `TensorBoardLogger` the definition could be something like `class TensorBoardLogger(SummaryWriterOptions, LightningLoggerBase):`.

    - Advantages: All Pitch features available. No need for deprecations or change how the classes are used.
    - Disadvantages: A bit more complex solution and feels a bit more hacky (though might not be).

5. Extend the support of `**kwargs` parameter discovery in jsonargparse using [ast](https://docs.python.org/3/library/ast.html), such that no refactoring is needed.

    - Advantages: All Pitch features available.
    - Disadvantages: Considerably complex and time consuming solution. Might not avoid the need of refactoring of all cases.

6. Please propose more possibilities.

The deprecations mentioned above could be optional. It wouldn't be a problem to keep the new parameter but also the existing `**kwargs`. This way there is not need for people to change their existing code.

The `dataclass_from_signature` function very likely will be added to and maintained in jsonargparse irrespective of this.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @justusschock @awaelchli @rohitgr7 @tchaton @carmocca @mauvilsa @akihironitta

### LINKED ISSUE #12697: [CLI] Can not instantiate DDPStrategy with find_unused_parameters=False using lazy_instance()
## 🐛 Bug

Instantiate `DDPStrategy` with `find_unused_parameters=False` raise exceptions when using `lazy_instance()`.
The non-lazy mode and no-parameter mode work as expected.

### To Reproduce

```python
import os

import torch
from torch.utils.data import DataLoader, Dataset

import pytorch_lightning as pl
from pytorch_lightning import LightningModule, Trainer
from pytorch_lightning.utilities import cli
from pytorch_lightning.strategies import DDPStrategy
from jsonargparse import lazy_instance


class RandomDataset(Dataset):
    def __init__(self, size, length):
        self.len = length
        self.data = torch.randn(length, size)

    def __getitem__(self, index):
        return self.data[index]

    def __len__(self):
        return self.len


class RandomDataModule(pl.LightningDataModule):
    def train_dataloader(self):
        return DataLoader(RandomDataset(32, 64), batch_size=2)


class BoringModel(LightningModule):
    def __init__(self):
        super().__init__()
        self.layer = torch.nn.Linear(32, 2)

    def forward(self, x):
        return self.layer(x)

    def training_step(self, batch, batch_idx):
        loss = self(batch).sum()
        self.log("train_loss", loss)
        return {"loss": loss}

    def configure_optimizers(self):
        return torch.optim.SGD(self.layer.parameters(), lr=0.1)


if __name__ == "__main__":
    c = cli.LightningCLI(
        BoringModel,
        RandomDataModule,
        trainer_defaults={
            # Case 1:
            # ValueError: 'Configuration check failed :: No action for
            # destination key "find_unused_parameters" to check its value.'
            'strategy': lazy_instance(DDPStrategy, find_unused_parameters=False),

            # Case 2: (lazy mode without find_unused_parameters)  work as expected
            'strategy': lazy_instance(DDPStrategy),

            # Case 3: (non-lazy mode) work as expected
            'strategy': DDPStrategy(find_unused_parameters=False),
        }
    )

```

### Expected behavior

All three cases are supposed to work.

### Environment

* CUDA:
	- GPU:
		- NVIDIA GeForce RTX 3090
	- available:         True
	- version:           11.3
* Packages:
	- numpy:             1.22.3
	- pyTorch_debug:     False
	- pyTorch_version:   1.11.0+cu113
	- pytorch-lightning: 1.6.0
	- tqdm:              4.64.0
* System:
	- OS:                Linux
	- architecture:
		- 64bit
		- ELF
	- processor:         x86_64
	- python:            3.9.11
	- version:           #44~20.04.1-Ubuntu SMP Thu Mar 24 16:43:35 UTC 2022

### PR DESCRIPTION
**Pull request by mauvilsa**
## What does this PR do?

Uses a new version of jsonargparse that extended the support for `**kwargs` to properly find signature parameters where possible by inspecting the source code, see https://github.com/omni-us/jsonargparse/issues/114#issuecomment-1152537388. The relevant documentation is at https://jsonargparse.readthedocs.io/en/latest/#parameter-resolvers.

Fixes #11653
Fixes #12697
Fixes #11574
Fixes #8561
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/9887#issuecomment-1018532746
Fixes https://github.com/PyTorchLightning/pytorch-lightning/issues/8008#issuecomment-945540010

### Does your PR introduce any breaking changes? If yes, please list them.

None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/utilities/test_cli.py [hunk 1]
```diff
@@ -1437,6 +1441,18 @@ def __init__(self, activation: torch.nn.Module = lazy_instance(torch.nn.LeakyReL
...
+def test_ddpstrategy_lazy_instance_with_find_unused_parameters():
+    with mock.patch("sys.argv", ["any.py"]):
+        cli = LightningCLI(
+            BoringModel,
+            run=False,
+            trainer_defaults={"strategy": lazy_instance(DDPStrategy, find_unused_parameters=False)},
+        )
+
+    assert cli.config.trainer.strategy.init_args.find_unused_parameters is False
+    assert isinstance(cli.config_init.trainer.strategy, DDPStrategy)
```
Comment (carmocca) [inline tests/utilities/test_cli.py:1453]: Can you add the same 2 checks but for the actual instance?

Comment (mauvilsa) [reply tests/utilities/test_cli.py]: It was checking the instance. Anyway I have changed the test checking more things.

**Comment by mauvilsa** [comment]
The jsonargparse version is 4.10.0.dev2 because from feedback here it might make sense to change things before a final release. Though the idea is that this pull request is merged only when 4.10.0 is out.

**Review by awaelchli — APPROVED** [review]

File: README.md [hunk 2]
```diff
@@ -36,14 +36,13 @@ ______________________________________________________________________
 
 ###### \*Codecov is > 90%+ but build delays may show less
 
-----
+______________________________________________________________________
```
Comment (carmocca) [inline README.md:39]: Are these accidental merge changes?

**Review by carmocca — APPROVED** [review]

File: README.md (hunk 2 above)
Comment (mauvilsa) [reply README.md:39]: pre-commit-ci[bot] did this and I don't know why, see https://github.com/Lightning-AI/lightning/pull/13283/commits/34fe10a8da95531f82489f96fd209a55e3095c84

**Review by Borda — APPROVED** [review]

**Pull request merged**